### PR TITLE
RBAC CLI Authorized Identity commands

### PR DESCRIPTION
**Pull request by peterschwarz**
This PR adds the `role` sub-commands `show`, `create`, `update`, `delete` to the CLI.

# Testing

### Build and configure your environment:

```
$ cargo build --manifest-path cli/Cargo.toml --features experimental
$ cargo build --manifest-path splinterd/Cargo.toml --features experimental
$ export PATH=target/debug:$PATH
$ export SPLINTER_HOME=/tmp/splinter 
```

These operations for setup can be done in one terminal.

First, initialize your splinter instance:

```
$ splinter cert generate
$ splinter database migrate
```

Also, generate a key and add it to the allow_keys file:
```
$ splinter keygen
$ cat ~/.cylinder/keys/$USER.pub >> /tmp/splinter/data/allow_keys
```

Start your splinter node:
```
$ splinterd --tls-insecure --enable-biome -vv --rest-api-endpoint http://127.0.0.1:8080
```

## Create Roles and authorized identities

These remainder of the operations in these test instructions can be done in a second terminal (once again, export your path from above). 

Add two role:

```
$  splinter role create \
    --display "Status Reader" \
    --permission "status.read" \
     status_reader
$  splinter role create \
    --display "Circuit Reader" \
    --permission "circuit.read" \
     circuit_reader
```

Show the currently authorized identities:

```
$ splinter authid list
IDENTITY                                                           TYPE ROLES
```

There are currently none, so lets create some:

```
$ splinter keygen -q test_user
$ splinter authid create --id-key $(cat ~/.cylinder/keys/test_user.pub) --role status_reader
```

Now list the keys again:
```
$ splinter authid list
IDENTITY                                                           TYPE ROLES
0261a18614bc108230310bd34f8c90ea9a65477b57feb292425f343dff279ec84c key  1
```

Check the authorized ID's details:

```
$ splinter authid show --id-key $(cat ~/.cylinder/keys/test_user.pub)
ID: 0261a18614bc108230310bd34f8c90ea9a65477b57feb292425f343dff279ec84c
    Type: key
    Roles:
        status_reader
```

Let's add the `circuit_reader` role, and show it again to verify the change

```
$ splinter authid update --id-key $(cat ~/.cylinder/keys/test_user.pub) --add-role circuit_reader
$ splinter authid show --id-key $(cat ~/.cylinder/keys/test_user.pub)
ID: 0261a18614bc108230310bd34f8c90ea9a65477b57feb292425f343dff279ec84c
    Type: key
    Roles:
        status_reader
        circuit_reader
```

Finally, let's delete the authorized identity and attempt to show the value agian

```
$ splinter authid delete --id-key $(cat ~/.cylinder/keys/test_user.pub)
$ splinter authid show --id-key $(cat ~/.cylinder/keys/test_user.pub)
ERROR: Subcommand encountered an error: Authorized identity key 0261a18614bc108230310bd34f8c90ea9a65477b57feb292425f343dff279ec84c does not exist
```

### CODE REVIEW

File: cli/man/splinter-authid-list.1.md [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+SYNOPSIS
+========
+**splinter authid list** \[**FLAGS**\] \[**OPTIONS**\]
+
+DESCRIPTION
+===========
+This command lists all of the authorized identities with assigned roles the
+local node has configured. This command displays abbreviated information
+pertaining to assignments in columns, with the headers `ID`, `TYPE`, and `#
+ROLES`. This makes quickly see which identities have assignments and to how many
```
Comment (ltseeley) [inline cli/man/splinter-authid-list.1.md:23]: Wording here needs to be fixed

File: cli/src/main.rs [hunk 2]
```diff
@@ -1070,6 +1070,53 @@ fn run<I: IntoIterator<Item = T>, T: Into<OsString> + Clone>(args: I) -> Result<
...
+                                .long("key")
+                                .takes_value(true)
+                                .help("Name or path of private key"),
+                        )
+                        .arg(
+                            Arg::with_name("id_key")
+                                .value_name("public-key")
+                                .long("id-key")
+                                .takes_value(true)
+                                .required(true)
```
Comment (ltseeley) [inline cli/src/main.rs:1142]: These should probably be `required_unless(<other_id_arg>)`

File: cli/man/splinter-authid-create.1.md [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+<!--
+  Copyright 2018-2021 Cargill Incorporated
+  Licensed under Creative Commons Attribution 4.0 International License
+  https://creativecommons.org/licenses/by/4.0/
+-->
+
+NAME
+====
+
+**splinter-authid-create** — Creates authorized identity on a Splinter node
```
Comment (ltseeley) [inline cli/man/splinter-authid-create.1.md:11]: s/Creates/Creates an/

File: cli/man/splinter-authid-create.1.md [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+: Specifies the private signing key (either a file path or the name of a
+  .priv file in $HOME/.splinter/keys).
+
+`-U`, `--url` URL
+: Specifies the URL for the `splinterd` REST API. The URL is required unless
+  `$SPLINTER_REST_API_URL` is set.
+
+`-D`, `--display` DISPLAY-NAME
+: Specifies the display name for the created role. This is a required option.
+
```
Comment (ltseeley) [inline cli/man/splinter-authid-create.1.md:52]: Copypasta

File: cli/man/splinter-authid-create.1.md [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+: Specifies a role to be included in the assignment. Specify multiple times for
+  more roles. At least one role is required.
+
+EXAMPLES
+========
+This example creates an authorized identity with two assigned roles.
+
+* The identity has user ID `"user-1234-abcd"`
+* There exists two roles on the system: `circuit_reader` and `status_reader`
+
```
Comment (ltseeley) [inline cli/man/splinter-authid-create.1.md:70]: There's an extra blank line here

File: cli/src/action/api/rbac/assignments.rs [hunk 6]
```diff
@@ -204,6 +245,53 @@ pub fn get_assignment(
...
+                        CliError::ActionError(format!(
+                            "Update assignment request failed with status code '{}', but error response \
+                            was not valid",
+                            status
+                        ))
+                    })?
+                    .message;
+
+                Err(CliError::ActionError(format!(
+                    "Update to create assignment: {}",
```
Comment (ltseeley) [inline cli/src/action/api/rbac/assignments.rs:288]: Should this be "Failed to update assignment"?

File: cli/src/main.rs [hunk 7]
```diff
@@ -1163,6 +1163,77 @@ fn run<I: IntoIterator<Item = T>, T: Into<OsString> + Clone>(args: I) -> Result<
...
+                        )
+                        .arg(
+                            Arg::with_name("rm_role")
+                                .value_name("role")
+                                .long("rm-role")
+                                .takes_value(true)
+                                .multiple(true)
+                                .number_of_values(1)
+                                .conflicts_with("rm_all")
+                                .help("A role to be added to the provided identity's assignments"),
```
Comment (ltseeley) [inline cli/src/main.rs:1225]: s/added to/removed from/

File: cli/src/main.rs [hunk 8]
```diff
@@ -1163,6 +1163,77 @@ fn run<I: IntoIterator<Item = T>, T: Into<OsString> + Clone>(args: I) -> Result<
...
+                        )
+                        .arg(
+                            Arg::with_name("force")
+                                .short("f")
+                                .long("force")
+                                .help("Ignore errors, such as adding and removing the same value."),
+                        )
+                        .arg(
+                            Arg::with_name("rm_role")
+                                .value_name("role")
```
Comment (ltseeley) [inline cli/src/main.rs:1219]: Is this `value_name("role")` intentional? Looks like it's the same as the one for `add-role`.

File: cli/man/splinter-authid-update.1.md [hunk 9]
```diff
@@ -0,0 +1,113 @@
...
+`-k`, `--key` PRIVATE-KEY-FILE
+: Specifies the private signing key (either a file path or the name of a
+  .priv file in $HOME/.splinter/keys).
+
+`-U`, `--url` URL
+: Specifies the URL for the `splinterd` REST API. The URL is required unless
+  `$SPLINTER_REST_API_URL` is set.
+
+`-D`, `--display` DISPLAY-NAME
+: Specifies the display name for the created role.
```
Comment (ltseeley) [inline cli/man/splinter-authid-update.1.md:56]: Copypasta

File: cli/man/splinter-authid-update.1.md [hunk 10]
```diff
@@ -0,0 +1,113 @@
...
+
+`--id-user` USER-ID
+: Specifies the user identity to update. Mutually exclusive to `--id-key`
+
+`--id-key` PUBLIC-KEY
+: Specifies the public key identity to update. Mutually exclusive to `--id-user`
+
+`--add-role` ROLE-ID 
+: Specifies a role to be add to the authorized identity. Specify multiple times
```
Comment (ltseeley) [inline cli/man/splinter-authid-update.1.md:65]: s/add/added/

File: cli/man/splinter-authid-update.1.md [hunk 11]
```diff
@@ -0,0 +1,113 @@
...
+
+`--rm-role` ROLE-ID
+: Specifies a role to be removed from the authorized identity. Specify multiple
+  times for more roles.
+
+
+EXAMPLES
+========
+This example updates an authorized identity with two assigned roles to include a
+third role.
```
Comment (ltseeley) [inline cli/man/splinter-authid-update.1.md:76]: This doesn't describe the role removal in the command

File: cli/man/splinter-authid.1.md [hunk 12]
```diff
@@ -48,11 +48,14 @@ SUBCOMMANDS
 : Shows a specific authorized identity on a Splinter node
 `create`
```
Comment (ltseeley) [inline cli/man/splinter-authid.1.md:49]: These should be alphabetical

File: cli/man/splinter-authid.1.md [hunk 13]
```diff
@@ -48,11 +48,14 @@ SUBCOMMANDS
...
 : Creates an authorized identity on a Splinter node.
+`update`
+: Updates an existing authorized identity on a Splinter node.
 
 SEE ALSO
 ========
 | `splinter-authid-list(1)`
 | `splinter-authid-show(1)`
 | `splinter-authid-create(1)`
```
Comment (ltseeley) [inline cli/man/splinter-authid.1.md:60]: These should be alphabetical

File: cli/src/action/api/rbac/assignments.rs [hunk 14]
```diff
@@ -292,6 +292,48 @@ pub fn update_assignment(
...
```
Comment (ltseeley) [inline cli/src/action/api/rbac/assignments.rs:330]: Should be "Failed to remove assignment"

File: cli/man/splinter-authid-delete.1.md [hunk 15]
```diff
@@ -0,0 +1,82 @@
...
+: Specifies the private signing key (either a file path or the name of a
+  .priv file in $HOME/.splinter/keys).
+
+`-U`, `--url` URL
+: Specifies the URL for the `splinterd` REST API. The URL is required unless
+  `$SPLINTER_REST_API_URL` is set.
+
+`-D`, `--display` DISPLAY-NAME
+: Specifies the display name for the created role.
+
```
Comment (ltseeley) [inline cli/man/splinter-authid-delete.1.md:50]: Copy pasta

File: cli/man/splinter-authid-delete.1.md [hunk 16]
```diff
@@ -0,0 +1,82 @@
...
+`--id-user` USER-ID
+: Specifies the user identity to delete. Mutually exclusive to `--id-key`
+
+`--id-key` PUBLIC-KEY
+: Specifies the public key identity to delete. Mutually exclusive to `--id-user`
+
+EXAMPLES
+========
+This example deletes an authorized identity with two assigned roles to include a
+third role.
```
Comment (ltseeley) [inline cli/man/splinter-authid-delete.1.md:60]: Copy pasta

File: cli/man/splinter-authid-delete.1.md [hunk 17]
```diff
@@ -0,0 +1,82 @@
...
+list the identity.
+
+ENVIRONMENT VARIABLES
+=====================
+**SPLINTER_REST_API_URL**
+: URL for the `splinterd` REST API. (See `-U`, `--url`.)
+
+SEE ALSO
+========
+| `splinter-authid-list(1)`
```
Comment (ltseeley) [inline cli/man/splinter-authid-delete.1.md:80]: These "SEE ALSO" sections should probably reference all other authid commands.

**Review by ltseeley — CHANGES_REQUESTED** [review]

**Comment by vaporos** [comment]
I found the "TYPE # ROLES" header confusing initially as it could potentially be three separate columns. "# ROLES" is also not friendly for parsing. Maybe just switch that to "ROLES"?

File: cli/src/main.rs (hunk 8 above)
Comment (peterschwarz) [reply cli/src/main.rs:1219]: This is the value displayed in the help: `--rm-role <role>`

File: cli/man/splinter-authid-delete.1.md (hunk 17 above)
Comment (peterschwarz) [reply cli/man/splinter-authid-delete.1.md:80]: I was assuming it was just the commands referenced in the example, but I suppose that makes sense

File: cli/man/splinter-authid.1.md (hunk 13 above)
Comment (peterschwarz) [reply cli/man/splinter-authid.1.md]: fixed; moved to the first commit where this file was introduced

File: cli/src/action/api/rbac/assignments.rs (hunk 14 above)
Comment (peterschwarz) [reply cli/src/action/api/rbac/assignments.rs]: fixed

File: cli/man/splinter-authid-delete.1.md (hunk 15 above)
Comment (peterschwarz) [reply cli/man/splinter-authid-delete.1.md]: fixed

File: cli/man/splinter-authid-delete.1.md (hunk 16 above)
Comment (peterschwarz) [reply cli/man/splinter-authid-delete.1.md:57]: fixed

File: cli/man/splinter-authid-create.1.md (hunk 5 above)
Comment (peterschwarz) [reply cli/man/splinter-authid-create.1.md:67]: fixed

File: cli/man/splinter-authid.1.md (hunk 12 above)
Comment (peterschwarz) [reply cli/man/splinter-authid.1.md:54]: Fixed - moved to first commit with this file

File: cli/man/splinter-authid-update.1.md (hunk 11 above)
Comment (peterschwarz) [reply cli/man/splinter-authid-update.1.md]: fixed

File: cli/man/splinter-authid-update.1.md (hunk 10 above)
Comment (peterschwarz) [reply cli/man/splinter-authid-update.1.md]: fixed

File: cli/man/splinter-authid-update.1.md (hunk 9 above)
Comment (peterschwarz) [reply cli/man/splinter-authid-update.1.md]: fixed

File: cli/src/main.rs (hunk 7 above)
Comment (peterschwarz) [reply cli/src/main.rs]: fixed

File: cli/src/action/api/rbac/assignments.rs (hunk 6 above)
Comment (peterschwarz) [reply cli/src/action/api/rbac/assignments.rs]: fixed

File: cli/man/splinter-authid-create.1.md (hunk 4 above)
Comment (peterschwarz) [reply cli/man/splinter-authid-create.1.md]: Fixed

File: cli/man/splinter-authid-create.1.md (hunk 3 above)
Comment (peterschwarz) [reply cli/man/splinter-authid-create.1.md]: Fixed

File: cli/src/main.rs (hunk 2 above)
Comment (peterschwarz) [reply cli/src/main.rs]: Fixed, and where each command is

File: cli/man/splinter-authid-list.1.md (hunk 1 above)
Comment (peterschwarz) [reply cli/man/splinter-authid-list.1.md]: fixed

**Review by vaporos — DISMISSED** [review]

File: cli/man/splinter-authid-delete.1.md (hunk 16 above)
Comment (ltseeley) [reply cli/man/splinter-authid-delete.1.md:57]: Looks like this wasn't changed

Comment (peterschwarz) [reply cli/man/splinter-authid-delete.1.md]: Fixed

**Review by ltseeley — APPROVED** [review]

**Review by vaporos — APPROVED** [review]

**Pull request merged**